torch/distributed/nn/functional.py

```diff
@@ -262,31 +264,38 @@ class _Reduce_Scatter(Function):
     @staticmethod
     def forward(ctx, op, group, tensor, *input_tensor_list):
         ctx.group = group
+        input_tensor_list = tuple(t.contiguous() for t in input_tensor_list)
         dist.reduce_scatter(tensor, list(input_tensor_list), op=op, group=group)
         return tensor
 
     @staticmethod
     def backward(ctx, grad_output):
-        return (None, None, None) + _AllGather.apply(
-            ctx.group, grad_output.contiguous()
-        )
+        return (None, None, None) + _AllGather.apply(ctx.group, grad_output)
 
 
 class _AllGather(Function):
     @staticmethod
     def forward(ctx, group, tensor):
         ctx.group = group
         out_tensor_list = [
-            torch.empty_like(tensor) for i in range(dist.get_world_size(group=group))
+            torch.empty_like(tensor) for _ in range(dist.get_world_size(group=group))
         ]
-        dist.all_gather(out_tensor_list, tensor, group=group)
+
+        dist.all_gather(out_tensor_list, tensor.contiguous(), group=group)
         return tuple(out_tensor_list)
 
     @staticmethod
     def backward(ctx, *grad_outputs):
-        tensor_list = [torch.empty_like(tensor) for tensor in grad_outputs]
-        gxs = _AlltoAll.apply(ctx.group, tensor_list, *grad_outputs)
-        gx = torch.sum(torch.stack(gxs), dim=0)
+        if dist.get_backend(group=ctx.group) is dist.Backend.GLOO:
+            # As gloo backend doesn't support ReduceScatter, we use AlltoAll with .sum()
+            # to emulate the ReduceScatter behavior
+            tensor_list = [torch.empty_like(tensor) for tensor in grad_outputs]
+            gxs = _AlltoAll.apply(ctx.group, tensor_list, *grad_outputs)
+            gx = torch.sum(torch.stack(gxs), dim=0)
+        else:
+            rank = dist.get_rank()
+            gx = torch.empty_like(grad_outputs[rank])
+            _Reduce_Scatter.apply(dist.ReduceOp.SUM, ctx.group, gx, *grad_outputs)
```

Comment:
I would opt to keep existing alltoall implementation for BC since many users have defined their own PG implementation.

Reply:
As discussed offline, update to use ReduceScatter only for NCCL backend